File: csharp/protein-translation/ProteinTranslation.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class ProteinTranslation
{
    private static (string[],string)[] codonMap = new[]
    {
        (new[] {"AUG"}, "Methionine"),
        (new[] {"UUU", "UUC"}, "Phenylalanine"),
        (new[] {"UUA", "UUG"}, "Leucine"),
        (new[] {"UCU", "UCC", "UCA", "UCG"}, "Serine"),
        (new[] {"UAU", "UAC"}, "Tyrosine"),
        (new[] {"UGU", "UGC"}, "Cysteine"),
        (new[] {"UGG"}, "Tryptophan"),
        (new[] {"UAA", "UAG", "UGA"}, null)
    };

    public static string[] Proteins(string strand)
        => Enumerable.Range(0, strand.Length / 3)
            .Select(o => strand.Substring(o * 3, 3))
            .Aggregate((Enumerable.Empty<string>(), false), (o, c) => 
            {
                var (set, done) = o;
                if (done) return o;

                var amino = codonMap.Single(map => map.Item1.Contains(c)).Item2;
                return amino == null ? (set, true) : (set.Append(amino), false);
            })
            .Item1.ToArray();
}

File: go/protein-translation/protein_translation.go
package protein

import "errors"

// STOP indicates a stop protein has been encountered
var STOP = errors.New("Processing terminated")

// ErrInvalidBase indicates a base has been found that is not recognised
var ErrInvalidBase = errors.New("Unrecognised base")

// FromCodon returns the appropriate protein from a given codon
func FromCodon(input string) (protein string, error error) {
	switch input {
	case "AUG":
		protein = "Methionine"
	case "UUU", "UUC":
		protein = "Phenylalanine"
	case "UUA", "UUG":
		protein = "Leucine"
	case "UCU", "UCC", "UCA", "UCG":
		protein = "Serine"
	case "UAU", "UAC":
		protein = "Tyrosine"
	case "UGU", "UGC":
		protein = "Cysteine"
	case "UGG":
		protein = "Tryptophan"
	case "UAA", "UAG", "UGA":
		error = STOP
	default:
		error = ErrInvalidBase
	}
	return
}

// FromRNA processes an RNA string and returns a set of proteins
func FromRNA(input string) (result []string, error error) {
	for i := 0; i < len(input); i += 3 {
		codon := input[i : i+3]
		protein, codonError := FromCodon(codon)
		if codonError == STOP {
			return
		} else if codonError != nil {
			error = codonError
			return
		}
		result = append(result, protein)
	}
	return
}
